server/etcdserver/version/version.go

```diff
@@ -29,27 +29,39 @@ var (
 	V3_6 = semver.Version{Major: 3, Minor: 6}
 )
 
-// UpdateStorageVersion updates storage version.
-func UpdateStorageVersion(lg *zap.Logger, tx backend.BatchTx) error {
-	tx.Lock()
-	defer tx.Unlock()
-	v, err := detectStorageVersion(lg, tx)
+// Migrate updates storage version to provided target version.
+func Migrate(lg *zap.Logger, tx backend.BatchTx, target semver.Version) error {
+	ver, err := detectStorageVersion(lg, tx)
 	if err != nil {
 		return fmt.Errorf("cannot determine storage version: %w", err)
 	}
-	switch *v {
-	case V3_5:
-		lg.Warn("setting storage version", zap.String("storage-version", V3_6.String()))
-		// All meta keys introduced in v3.6 should be filled in here.
-		buckets.UnsafeSetStorageVersion(tx, &V3_6)
-	case V3_6:
-	default:
-		lg.Warn("unknown storage version", zap.String("storage-version", v.String()))
+	if ver.Major != target.Major {
+		lg.Panic("Changing major storage version is not supported",
```

Comment:
Just a thought, what will happen in v4, how will downgrades/migrates work then?

Reply:
We will update this code when we decide to release v4.0.0. If for example v3.7 will be the last v3.X release then we will add a code in v4.0.0 that will treat v3.7 as predecessor. In that situation migration capabilities should be:
* etcd v3.6 can run on data from: v3.5, v3.6
* etcdutl v3.6 can migrate storage between versions: v3.5, v3.6
* etcd v3.7 can run on data from: v3.6, v3.7
* etcdutl v3.7 can migrate storage between versions: v3.5, v3.6, v3.7
* etcd v4.0 can run on data from: v3.7, v4.0
* etcdutl v4.0 can migrate storage between versions: v3.5, v3.6, v3.7, v4.0

In short etcd will work with data from it's or previous release. However if you have etcdutl from latest release you should be able to upgrade and downgrade storage freely.

Reply:
Thank you for the details!